File: easy/121.best-time-to-buy-and-sell-stock.py
#
# @lc app=leetcode id=121 lang=python3
#
# [121] Best Time to Buy and Sell Stock
#

# @lc code=start
class Solution:
    # Time Complexity is O(n)
    def maxProfit(self, prices: List[int]) -> int:
        mx = 0
        mn = prices[0]

        for price in prices:
            mn = min(mn, price)
            profit = price - mn
            mx = max(mx, profit)
        return mx
# @lc code=end

File: easy/121.best-time-to-buy-and-sell-stock.go
package code

/*
 * @lc app=leetcode id=121 lang=golang
 *
 * [121] Best Time to Buy and Sell Stock
 */

// @lc code=start
func maxProfit(prices []int) int {
	max := 0
	min := prices[0]

	profit := 0

	for _, price := range prices {
		if price < min {
			min = price
		}
		profit = price - min
		if profit > max {
			max = profit
		}
	}
	return max
}

// @lc code=end
